File: 2025/42. Trapping Rain Water/Solution2.java
class Solution {
    public int trap(int[] height) {
        int left = 0, right = height.length - 1;
        int maxLeft = height[left], maxRight = height[right];
        int sum = 0;
        while (left <= right) {
            if (maxLeft <= maxRight) {
                maxLeft = Math.max(height[left], maxLeft);
                sum += maxLeft - height[left];
                left ++;
            } else {
                maxRight = Math.max(height[right], maxRight);
                sum += maxRight - height[right];
                right --;
            }
        }
        return sum;
    }
}

File: 2025/122. Best Time to Buy and Sell Stock II/solution2.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        dp = [[0] * 2 for _ in range(n + 1)]
        for i in range(n - 1, -1, -1):
            for j in range(2):
                if j == 0:
                    dp[i][j] = dp[i + 1][1] - prices[i]
                else:
                    dp[i][j] = dp[i + 1][0] + prices[i]
                dp[i][j] = max(dp[i + 1][j], dp[i][j])
        return dp[0][0]